### PR TITLE
Upgrade to Rails 4.2 

### PR DESCRIPTION
**Pull request by Souravirus**
Related to #1832 
Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] PR body includes `fixes #0000`-style reference to original issue #
* [x] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software

We have a loose schedule of reviewing and pulling in changes every Tuesday and Friday, and publishing changes on Fridays. 

Thanks!

### CODE REVIEW

**Comment by jywarren** [comment]
Connecting with https://github.com/publiclab/plots2/pull/1832 -- thanks @Souravirus !!!

**Comment by jywarren** [comment]
Hmm, very close, it seems? Wow!

File: app/controllers/notes_controller.rb [hunk 1]
```diff
@@ -111,14 +111,11 @@ def create
           @node.add_tag('event:rsvp', current_user)
           @node.add_tag('date:' + params[:date], current_user) if params[:date]
         end
-        if params[:draft] != true
```
Comment (jywarren) [inline app/controllers/notes_controller.rb]: Oops, are we losing the draft feature here?

File: app/controllers/user_sessions_controller.rb [hunk 2]
```diff
@@ -25,7 +25,7 @@ def create
           params[:user_session].delete(:username)
           params[:openid] = username # pack up username for second runthrough
         end
-        @user_session = UserSession.new(params[:user_session])
+        @user_session = UserSession.new(username: params[:user_session][:username], password: params[:user_session][:password], remember_me: params[:user_session][:remember_me])
```
Comment (jywarren) [inline app/controllers/user_sessions_controller.rb]: Maybe on separate lines?

File: app/mailers/subscription_mailer.rb [hunk 3]
```diff
@@ -41,7 +41,7 @@ def notify_tag_added(node, tag, current_user)
     final_users_to_email.each do |user|
       @user = user
       unless user.id == current_user.id 
-         mail(to: user.email, subject: "Added: #{node.title}").deliver
+         mail(to: user.email, subject: "New tag added on #{node.title}")
```
Comment (jywarren) [inline app/mailers/subscription_mailer.rb]: Ah, we'd meant to change this but it looks reverted?

File: app/controllers/notes_controller.rb (hunk 1 above)
Comment (Souravirus) [reply app/controllers/notes_controller.rb]: Ohh I mistakenly deleted it, I will add it shortly

File: app/controllers/user_sessions_controller.rb (hunk 2 above)
Comment (Souravirus) [reply app/controllers/user_sessions_controller.rb]: Ohh sorry! I will indent it

File: app/mailers/subscription_mailer.rb (hunk 3 above)
Comment (Souravirus) [reply app/mailers/subscription_mailer.rb]: Sure I will change it. Thanks!!

**Comment by Souravirus** [comment]
Thanks for review @jywarren

**Comment by jywarren** [comment]
Keep up the good work! I see you making progress on this and it's exciting!

**Comment by Souravirus** [comment]
Hey @jywarren, why the travis test is not running whereas its running on my fork. Here is the pic. 
screenshot from 2018-04-01 12-57-06

**Comment by Souravirus** [comment]
It worked by by reopening the pull request

**Comment by Souravirus** [comment]
Hey @jywarren, I have finally got the green tick, I mean I removed all the errors and fails. Please review it. I now have to remove the depreciation warning and other warnings. Please see this. Thanks !!

**Comment by jywarren** [comment]
OH MY GOSH - wow, you did it? Looking now. @publiclab/reviewers please check out this PR and look through it to ensure no recent code additions have been reverted, as it's a very big rebase. 

MANY kudos to @Souravirus for taking on this very big challenge and sticking with it for several weeks!

📯 🎉 🎂 🕺 👍 👍

**Comment by jywarren** [comment]
This looks pretty good to me, to be honest. Almost all changes are minor OR changes to `deliver` and `deliver_now`. Fantastic work!

File: test/integration/node_unique_views_test.rb [hunk 4]
```diff
@@ -7,14 +7,14 @@ class NodeInsertExtrasTest < ActionDispatch::IntegrationTest
     assert_equal 0, Impression.count
 
     assert_difference 'Impression.count', 1 do
-      get "wiki/#{nodes(:about).slug}"
+      get "/wiki/#{nodes(:about).slug}" #gave the complete address changed from wiki to /wiki
```
Comment (Gauravano) [inline test/integration/node_unique_views_test.rb]: feel free to remove this comment anytime

File: test/integration/node_unique_views_test.rb [hunk 5]
```diff
@@ -7,14 +7,14 @@ class NodeInsertExtrasTest < ActionDispatch::IntegrationTest
...
 
       assert_response :success
     end
 
     assert_difference 'nodes(:about).totalviews', 0 do
       assert_difference 'Impression.count', 0 do
-        get "wiki/#{nodes(:about).slug}"
+        get "/wiki/#{nodes(:about).slug}" #gave the complete address changed from wiki to /wiki
```
Comment (Gauravano) [inline test/integration/node_unique_views_test.rb]: this comment can also be removed

File: config/database.yml.example [hunk 6]
```diff
@@ -10,6 +10,7 @@ test:
   encoding: utf8
   pool: 5
   timeout: 5000
+  strict: false
```
Comment (ryzokuken) [inline config/database.yml.example]: Same as above

File: config/database.yml.example [hunk 7]
```diff
@@ -18,6 +19,7 @@ development:
   password: 
   encoding: utf8
   database: publiclaboratory
+  strict: false
```
Comment (ryzokuken) [inline config/database.yml.example]: Same as above

File: test/integration/node_unique_views_test.rb (hunk 4 above)
Comment (Souravirus) [reply test/integration/node_unique_views_test.rb]: Yeah @Gauravano I will remove these

File: config/database.yml.example [hunk 8]
```diff
@@ -28,3 +30,4 @@ production:
   encoding: utf8
   pool: 5
   timeout: 5000
+  strict: false
```
Comment (ryzokuken) [inline config/database.yml.example]: Same as above

File: config/environments/production.rb [hunk 9]
```diff
@@ -33,11 +33,12 @@
...
   # Force all access to the app over SSL, use Strict-Transport-Security, and use secure cookies.
-  #config.force_ssl = true
+  # config.force_ssl = true
   # config.middleware.insert_before ActionDispatch::Static, Rack::SSL, :exclude => proc { |env| env['HTTPS'] != 'on' }
 
-  # See everything in the log (default is :info)
-  # config.log_level = :debug
+  # Set to `:info` to match the current default, or set to `:debug` to opt-into
+  # the future default.
+  config.log_level = :info
```
Comment (ryzokuken) [inline config/environments/production.rb]: Same as above

File: app/controllers/answers_controller.rb [hunk 10]
```diff
@@ -65,8 +65,7 @@ def accept
         else
           @answer.accepted = true
           @answer.save
-          @answer.node.add_tag('answered', @answer.author)
```
[the remark this answers is not in the document]
Comment (Souravirus) [reply app/controllers/answers_controller.rb]: I removed it mistakenly. I would add it again. Thanks!!

File: app/models/drupal_content_type_map.rb [hunk 11]
```diff
@@ -12,6 +12,12 @@ class DrupalContentTypeMap < ActiveRecord::Base
                    lat_column_name: :lat,
                    lng_column_name: :lng
 
+  before_save :truncate_fields
+
+  def truncate_fields
```
[the remark this answers is not in the document]
Comment (Souravirus) [reply app/models/drupal_content_type_map.rb]: screenshot from 2018-04-06 22-13-24
Actually these were added by @jywarren during the rails 4.2 update

[the remark this answers is not in the document]
Comment (jywarren) [reply app/models/drupal_content_type_map.rb]: Early in the upgrade process I found I had to do this to get one test to pass, so I believe for some reason this was 4.2 related. Good question!

[the remark this answers is not in the document]
Comment (ryzokuken) [reply app/models/drupal_content_type_map.rb]: Okay. @jywarren what do you say? I'd rather do a single thing in each PR.

**Comment by ryzokuken** [comment]
Also, when we merge this, please make sure to squash and merge this.

@Souravirus if possible, once the contents of this PR get approved completely, squash them.

File: config/application.rb [hunk 12]
```diff
@@ -106,5 +106,6 @@ class Application < Rails::Application
 
     # Allow mass assignments
     config.active_record.whitelist_attributes = false
+    config.active_record.raise_in_transactional_callbacks = true
```
[the remark this answers is not in the document]
Comment (Souravirus) [reply config/application.rb]: No @ryzokuken, its a commit regarding to the update 
screenshot from 2018-04-06 22-31-54

File: config/database.yml.example (hunk 6 above)
Comment (Souravirus) [reply config/database.yml.example]: This was added by @jywarren in the update process.

File: config/database.yml.example (hunk 7 above)
Comment (Souravirus) [reply config/database.yml.example]: This was added by @jywarren in the update process

File: config/database.yml.example (hunk 8 above)
Comment (Souravirus) [reply config/database.yml.example]: This was added by @jywarren in the update process

File: config/environments/production.rb (hunk 9 above)
Comment (Souravirus) [reply config/environments/production.rb]: This was added by @jywarren in the update process

**Comment by Souravirus** [comment]
Yeah definitely I will squash them thanks @ryzokuken

File: config/database.yml.example (hunk 6 above)
Comment (jywarren) [reply config/database.yml.example]: Again, this was to get a test to pass. Thanks!

**Comment by Souravirus** [comment]
Done the changes as suggested by @Gauravano  and @ryzokuken . Please see these changes @Gauravano, @ryzokuken @jywarren. Thanks!!

File: app/models/drupal_content_type_map.rb (hunk 11 above)
[the remark this answers is not in the document]
Comment (jywarren) [reply app/models/drupal_content_type_map.rb]: The ones I added were Rails 4.2 specific, so I'm fine with them being squashed into this. Sometimes with our atypical legacy code we require workarounds to move up Rails versions which may not immediately be apparent from the upgrade checklist but are borne out by the CI tests. 

More broadly I am very much in favor of what you're saying -- if things can be merged in separately, then we can refer back to this PR as a good example of a version upgrade! In this case, as soon as @Souravirus re-adds the line on line 68 of `answers_controller.rb`, I think we are good to move forward. Thanks for such a careful read-through!

**Comment by jywarren** [comment]
OK -- perfect. I will check with @icarito in a few hours for one final check and we will move ahead with this!

**Comment by Souravirus** [comment]
Okk thanks

Sent from Yahoo Mail on Android 
 
  On Fri, Apr 6, 2018 at 23:16, Jeffrey Warren<notifications@github.com> wrote:   
OK -- perfect. I will check with @icarito in a few hours for one final check and we will move ahead with this!

**Comment by Souravirus** [comment]
Okk thanks @jywarren

**Comment by jywarren** [comment]
This is merged now!!

**Comment by jywarren** [comment]
Congratulations @Souravirus and everybody!

**Comment by Souravirus** [comment]
Thanks !!

**Comment by Souravirus** [comment]
@jywarren when it will be updated in the original website?

**Comment by seafr** [comment]
@Souravirus congrats on successfully taking on such a big challenge.

**Comment by Souravirus** [comment]
thanks @seafr

**Comment by icarito** [comment]
Hey great work here, it's live on the publiclab.org website! Thanks for your contribution!

**Comment by Souravirus** [comment]
Ah! great. Thanks!!

**Pull request merged**